text/0000-pack-subcommands.md

```diff
@@ -0,0 +1,186 @@
+  completion            Outputs completion script location
+  help                  Help about any command
+```
+
+# What it is
+[what-it-is]: #what-it-is
+
+
+### Proposed structure
+
+```
+pack
+
+  app
+    build
+    inspect
+    rebase
+  builder
+    create
+    inspect
+    suggest
+  buildpack
+    package
+    register
+  completion
+  config
+    default-builder
+      set
+      unset
+      show
+    run-image-mirror
+      add
+      remove
+      list
+    registry
+      add
+      remove
+      list
+    trusted-builder
+      add
+      remove
+      list
+  help
+  report
+  stack
+    suggest
+  version
+```
+
+### Top-Level Aliases
+
+There would also be some alias subcommands for the `app` set of subcommands for easier usage and compatibility.
+
+* `pack build` -> `pack app build`
+* `pack inspect` -> `pack app inspect`
+* `pack rebase` -> `pack app rebase`
+
+### Config
+
+The `config` subcommand would work with the notion of two different type of resources:
+
+##### Singular
+
+Singular resources would have the following subcommands:
+
+* set
+* unset
+* show (aliased to subcommand root)
+
+###### Example
+
+```
+# sets the default builder
+pack config default-builder set gcr.io/my-org/my-builder:latest
+
+# unset the default builder
+pack config default-builder unset
+
+# show the default builder
+pack config default-builder show
+
+# show the default builder (w/o subcommand)
+pack config default-builder
+```
+
+##### List
+
+List resources would have the following subcommands:
+
+* add
+* remove
+* list
+
+The explicit additional parameters would be based on the specific implementation needs.
+
+###### Example
+
+```
+# adds `gcr.io/my-org/my-image` as a mirror for `my-image`  
+pack config run-image-mirrors add my-image gcr.io/my-org/my-image
+
+# removes `gcr.io/my-org/my-image` as a mirror for `my-image`
+pack config run-image-mirrors remove my-image gcr.io/my-org/my-image
+
+# removes all mirrors for `my-image`
+pack config run-image-mirrors remove my-image
+
+# list all mirrors
+pack config run-image-mirrors list
+
+# list all mirrors for `my-image`
+pack config run-image-mirrors list my-image
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+
+## Backwards Compatibility
+
+1. There would be a period in which all pre-existing subcommands would work.
+2. [Top-level aliases](#top-level-aliases) will provide continued compatibility for common operations.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Change is bad... mkay.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Keep it as it is.
+- Some [Config](#config) subcommands could be more intertwined with their respective resource. For example, `pack builder trust ...` instead of `pack config trust-builder ...`.
+
+# Prior Art
+[prior-art]: #prior-art
+
+Discuss prior art, both the good and bad.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Are there better alternatives to the name "app"?
```

Comment:
image?

Reply:
I'm not keen on image because on the technical side almost every artifact we use is an image. Additionally, some args are named "image-name" which refer to OCI images in general and not the "resulting image produced from a pack build".

Reply:
I am +1 on "app"

In the spec we ended up going with "app image" everywhere for clarity. I know at one point we wanted to steer clear of "app" but I am now of the opinion that not using `app` just makes everything more confusing (even if some images aren't strictly speaking "apps").